packages/gatsby-source-filesystem/src/gatsby-node.js

```diff
@@ -11,218 +11,159 @@ const { createId, createFileNode } = require(`./create-file-node`)
-            CHOKIDAR_READY: "CHOKIDAR_WATCHING",
-            BOOTSTRAP_FINISHED: "CHOKIDAR_WATCHING_BOOTSTRAP_FINISHED",
-          },
-        },
-        CHOKIDAR_WATCHING: {
+        SRC_FS_BUSY: {
           on: {
-            BOOTSTRAP_FINISHED: "CHOKIDAR_WATCHING_BOOTSTRAP_FINISHED",
-            CHOKIDAR_READY: "CHOKIDAR_WATCHING",
+            SRC_FS_IDLE: `SRC_FS_IDLE`,
           },
         },
-        CHOKIDAR_WATCHING_BOOTSTRAP_FINISHED: {
+        SRC_FS_IDLE: {
           on: {
-            CHOKIDAR_READY: "CHOKIDAR_WATCHING_BOOTSTRAP_FINISHED",
+            SRC_FS_BUSY: `SRC_FS_BUSY`,
           },
         },
-      },
-    },
-    PROCESSING: {
-      initial: `BOOTSTRAPPING`,
-      states: {
-        BOOTSTRAPPING: {
+        SRC_FS_INIT: {
           on: {
-            BOOTSTRAP_FINISHED: "IDLE",
+            SRC_FS_READY: `SRC_FS_READY`,
+            SRC_FS_SETUP: `SRC_FS_SETUP`,
           },
         },
-        IDLE: {
+        SRC_FS_READY: {
           on: {
-            EMIT_FS_EVENT: `PROCESSING`,
+            SRC_FS_SETUP: `SRC_FS_IDLE`,
           },
         },
-        PROCESSING: {
+        SRC_FS_SETUP: {
           on: {
-            QUERY_QUEUE_DRAINED: `IDLE`,
-            TOUCH_NODE: `IDLE`,
+            SRC_FS_READY: `SRC_FS_IDLE`,
           },
         },
       },
     },
   },
+  strict: true,
 })
 
-let currentState = fsMachine.initialState
-
-const fileQueue = new Map()
+let state = fsMachine.initialState
 
 exports.sourceNodes = (
-  { actions, getNode, createNodeId, hasNodeChanged, reporter, emitter },
-  pluginOptions
+  { actions, createNodeId, emitter, getNode, reporter },
+  opts = {}
 ) => {
   const { createNode, deleteNode } = actions
 
-  // Validate that the path exists.
-  if (!fs.existsSync(pluginOptions.path)) {
-    reporter.panic(`
-The path passed to gatsby-source-filesystem does not exist on your file system:
-
-${pluginOptions.path}
-
-Please pick a path to an existing directory.
+  // Verify the path exists.
+  if (!fs.existsSync(opts.path)) {
+    reporter.panic(`gatsby-source-filesystem: "${opts.path}" does not exist.
+    Specify the path to an existing directory.
 
-See docs here - https://www.gatsbyjs.org/packages/gatsby-source-filesystem/
-      `)
+    Visit https://www.gatsbyjs.org/packages/gatsby-source-filesystem/ for details.
+    `)
   }
-  let fileNodeQueue = new Map()
 
-  // Once bootstrap is finished, we only let one File node update go through
-  // the system at a time.
-  emitter.on(`BOOTSTRAP_FINISHED`, () => {
-    currentState = fsMachine.transition(
-      currentState.value,
-      `BOOTSTRAP_FINISHED`
-    )
-  })
-  emitter.on(`TOUCH_NODE`, () => {
-    // If we create a node which is the same as the previous version, createNode
-    // returns TOUCH_NODE and then nothing else happens so we listen to that
-    // to return the state back to IDLE.
-    currentState = fsMachine.transition(currentState.value, `TOUCH_NODE`)
-  })
+  const queue = (() => {
```

Comment:
I'm not sure what your intention was with this queue?

Reply:
Instead of queues for different ready states, this queue is a single unified queue that all chokidar events can add to and if the system is ready to process them, just run them. Having one singleton queue vs multiple and doing the is-ready check inside the queue instead of at each event seems simpler.

Reply:
Perhaps this is the source of confusion — there wasn't any queues before — we're using xstate now which let's us create a statechart for the plugin — https://github.com/davidkpiano/xstate/

Reply:
The queues I was referring to are:
https://github.com/gatsbyjs/gatsby/blob/v2/packages/gatsby-source-filesystem/src/gatsby-node.js
lines: 71, 91, 114, 153.

This one function handles all of those cases in a consistent manner. Once chokidar is in a ready state, changes are added to the queue (for no other reason than simplicity and consistency) and immediately flushed.

Reply:
Ah right :-) There was that queue.

To simplify this PR — could you just back out all the changes to the queue and to the state machine?

Reply:
Done. (I think.)